src/bip322.rs

```diff
@@ -0,0 +1,337 @@
+    MIDSTATE,
+    64,
+    doc = "test hash",
+    true
+);
+
+/// BIP322 Error types
+#[derive(Debug)]
+pub enum BIP322Error {
+    /// BIP322 Internal Error
+    InternalError(String),
+
+    /// Signature Validation Error
+    ValidationError(InterpreterError),
+}
+
+#[doc(hidden)]
+impl From<InterpreterError> for BIP322Error {
+    fn from(e: InterpreterError) -> BIP322Error {
+        BIP322Error::ValidationError(e)
+    }
+}
+
+/// Bip322 Signatures
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Bip322Signature {
+    /// Legacy style. Only applicable for P2PKH message_challenge
+    Legacy(Signature, PublicKey),
+
+    /// Simple witness structure
+    Simple(Vec<Vec<u8>>),
+
+    /// Full `to_sign` transaction structure
+    Full(Transaction),
+}
+
+/// BIP322 validator structure
+/// A standard for interoperable signed messages based on the Bitcoin Script format,
+/// either for proving fund availability, or committing to a message as the intended
+/// recipient of funds sent to the invoice address.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Bip322<T: MiniscriptKey + ToPublicKey> {
+    /// Message to be signed
+    message: Vec<u8>,
+
+    /// Signature to verify the message
+    /// Optional value is used here because a validator structure can be
+    /// created without a BIP322Signature. Such structure can only produce
+    /// to_spend (or empty to_sign) transaction, but cannot validate them.
+    signature: Option<Bip322Signature>,
+
+    /// script_pubkey to define the challenge script inside to_spend transaction
+    /// here we take in descriptors to derive the resulting script_pubkey
+    message_challenge: Descriptor<T>,
+}
+
+impl<T: MiniscriptKey + ToPublicKey> Bip322<T> {
+    /// Create a new BIP322 validator
+    pub fn new(msg: &[u8], sig: Option<Bip322Signature>, addr: Descriptor<T>) -> Self {
+        Bip322 {
+            message: msg.to_vec(),
+            signature: sig,
+            message_challenge: addr,
+        }
+    }
+
+    /// Insert Signature inside BIP322 structure
+    pub fn insert_sig(&mut self, sig: Bip322Signature) {
+        self.signature = Some(sig)
+    }
+
+    /// create the to_spend transaction
+    pub fn to_spend(&self) -> Transaction {
+        // create default input and output
+        let mut vin = TxIn::default();
+        let mut vout = TxOut::default();
+
+        // calculate the message tagged hash
+        let msg_hash = MessageHash::hash(&self.message[..]).into_inner();
+
+        // mutate the input with appropriate script_sig and sequence
+        vin.script_sig = Builder::new()
+            .push_int(0)
+            .push_slice(&msg_hash[..])
+            .into_script();
+        vin.sequence = 0;
+
+        // mutate the value and script_pubkey as appropriate
+        vout.value = 0;
+        vout.script_pubkey = self.message_challenge.script_pubkey();
+
+        // create and return final transaction
+        Transaction {
+            version: 0,
```

Comment:
I have made a comment to BIP322 to make this always 2 and it was positively received. I think we should keep it always 2.

Reply:
Done..